tools/halide_image.h

```diff
@@ -1,278 +1,673 @@
+        *result = sizeof...(args);
     }
 
-    int min(int dim) const {
-        return contents->buf.min[dim];
+    /** The recursive version is only enabled up to a recursion limit
+     * of 256. This catches callables that aren't callable with any
+     * number of ints. */
+    template<typename ...Args>
+    static void num_args(double, int *result, Fn f, Args... args) {
+        static_assert(sizeof...(args) <= 256,
+                      "Callable passed to for_each_element must accept either a const int *,"
+                      " or up to 256 ints. No such operator found. Expect infinite template recursion.");
+        return num_args(0, result, f, 0, args...);
     }
 
-    int extent(int dim) const {
-        return contents->buf.extent[dim];
+    static int get_number_of_args(Fn f) {
+        int result;
+        num_args(0, &result, f);
+        return result;
     }
 
-    void set_min(int x, int y = 0, int z = 0, int w = 0) {
-        contents->buf.min[0] = x;
-        contents->buf.min[1] = y;
-        contents->buf.min[2] = z;
-        contents->buf.min[3] = w;
+    /** A run-time-recursive version (instead of
+     * compile-time-recursive) that requires the callable to take a
+     * pointer to a position array instead. */
+    static void for_each_element_array(int d, Fn f, const buffer_t &buf, int *pos) {
+        if (d == -1) {
+            f(pos);
+        } else {
+            pos[d] = buf.min[d];
+            for (int i = 0; i < std::max(1, buf.extent[d]); i++) {
+                for_each_element_array(d - 1, f, buf, pos);
+                pos[d]++;
+            }
+        }
+    }
+
+    /** We now have two overloads for for_each_element. This one
+     * triggers if the callable takes a const int *.
+     */
+    template<typename Fn2>
+    static auto for_each_element(int, const buffer_t &buf, Fn2 f)
+        -> decltype(f((const int *)0)) {
+        int pos[4] = {0, 0, 0, 0};
+        int dimensions = 0;
+        while (buf.extent[dimensions] != 0 && dimensions < 4) {
+            dimensions++;
+        }
+        for_each_element_array(dimensions - 1, f, buf, pos);
+    }
+
+    /** This one triggers otherwise. It treats the callable as
+     * something that takes some number of ints. */
+    template<typename Fn2>
+    static void for_each_element(double, const buffer_t &buf, Fn2 f) {
+        int num_args = get_number_of_args(f);
+        for_each_element_variadic(0, num_args-1, f, buf);
     }
 };
 
+/** Call a function at each site in a buffer. If the function has more
+ * arguments than the buffer has dimensions, the remaining arguments
+ * will be zero. If it has fewer arguments than the buffer has
+ * dimensions then the last few dimensions of the buffer are not
+ * iterated over. For example, the following code exploits this to set
+ * a floating point RGB image to red:
+
+\code
+Image<float, 3> im(100, 100, 3);
+for_each_element(im, [&](int x, int y) {
```

Comment:
Actually it's still useful. It's easy to get the bounds wrong on the nested for loop. I also improved the performance of the const int * version for generic code.